File: py_solutions_1-10/Euler_3.py
# what is the largest prime factor of 600851475143 ? 

from __future__ import print_function

import math
import timeit


try:
    range = xrange
except NameError:
    pass
 
start = timeit.default_timer()

def is_prime(x):
    if x == 2:
        return True
    if x == 1 or x % 2 == 0:
        return False
    for i in range(3,int(math.sqrt(x)) + 1, 2):
        if x % i == 0:
            return False
    return True

def euler_3():
    FACTOR_OF = 600851475143
    return max([i for i in range(1, int(math.sqrt(FACTOR_OF)) + 1, 2) 
                if FACTOR_OF % i == 0 and is_prime(i)])


print("Answer: {}".format(euler_3()))
stop = timeit.default_timer()
print("Time: {0:9.5f}".format(stop - start))

File: Rust/Euler_3.rs
// what is the largest prime factor of 600851475143 ? 

use std::time::Instant;

fn is_prime(n:u64) -> bool {
    if n == 2 {
        return true;
    }

    if n < 2 || n % 2 == 0 {
        return false;
    }

    let range:u64 = n.sqrt() + 1;

    for i in 3..range {
        if n % i == 0 {
            return false;
        }
    }

    return true;
}

fn main() {
    let timer = Instant::now();

    let factor:u64 = 600851475143;
    let current:u64 = factor / 2;

    let answer = {
        while true {
            if factor % current == 0 && is_prime(current) {
                break current;
            }

            current -= 1;
        };
    };

    println!("Answer: {}", answer);
    println!("Time: {:?}", timer.elapsed());
}
